snuba/query/query.py

```diff
@@ -204,3 +219,135 @@ def get_all_referenced_columns(self) -> Sequence[Any]:
 
         # Return the set of all columns referenced in any expression
         return set(chain(*[columns_in_expr(ex) for ex in col_exprs]))
+
+    def __replace_col_in_expression(
+        self,
+        expression: Any,
+        old_column: str,
+        new_column: str,
+    ) -> Any:
+        """
+        Returns a copy of the expression with old_column replaced by new_column.
+        This does not modify anything in place since expressions can be either
+        mutable (lists) or immutable (strings and tuples), so the only way to be
+        consistent in the result is always to return a brand new expression.
+
+        This does not support all possible columns format (like string representations
+        of funciton expresison "count(col1)"). But it is consistent with the behavior
+        of get_all_referenced_columns in that it won't replace something that is not
+        returned by that function and it will replace all the columns reteturned by that
+        function.
+        """
+        if isinstance(expression, str):
+            match = ORDERED_COL.match(expression)
+            if match and match[1] == old_column:
+                return expression.replace(old_column, new_column)
+        elif (isinstance(expression, (list, tuple)) and len(expression) >= 2
+                and isinstance(expression[1], (list, tuple))):
+            params = [
+                self.__replace_col_in_expression(param, old_column, new_column)
+                for param in expression[1]
+            ]
+            ret = [expression[0], params]
+            if len(expression) == 3:
+                # Alias for the column
+                ret.append(expression[2])
+            return ret
+        return expression
+
+    def __replace_col_in_condition(
+        self,
+        condition: Condition,
+        old_column: str,
+        new_column: str,
+    ) -> Condition:
+        """
+        Replaces a column in a structured condition. This is a level above replace_col_in_expression
+        since conditions are in the form [expression, operator, literal] (which is not fully correct
+        since the right side of the condition should be an expression as well but this constraint is
+        imposed by the query schema and get_all_referenced_columns behaves accordingly).
+        Conditions can also be nested.
+        """
+        if is_condition(condition):
+            return [
+                self.__replace_col_in_expression(condition[0], old_column, new_column),
+                condition[1],
+                condition[2]
+            ]
+        elif isinstance(condition, (tuple, list)):
+            # nested condition
+            return [
+                self.__replace_col_in_condition(cond, old_column, new_column) for cond in condition
+            ]
+        else:
+            # Don't know what this is
+            return condition
```

Comment:
I think there are some corner cases where this can happen from the query schema. Though I could not effectively pinpoint a specific one.